cpp/dolfinx/la/MatrixCSR.h

```diff
@@ -630,15 +630,17 @@ std::vector<typename MatrixCSR<U, V, W, X>::value_type>
 MatrixCSR<U, V, W, X>::to_dense() const
 {
   const std::size_t nrows = num_all_rows();
-  const std::size_t ncols
-      = _index_maps[1]->size_local() + _index_maps[1]->num_ghosts();
+  const std::size_t ncols = _index_maps[1]->size_global();
   std::vector<value_type> A(nrows * ncols * _bs[0] * _bs[1], 0.0);
   for (std::size_t r = 0; r < nrows; ++r)
     for (std::int32_t j = _row_ptr[r]; j < _row_ptr[r + 1]; ++j)
       for (int i0 = 0; i0 < _bs[0]; ++i0)
         for (int i1 = 0; i1 < _bs[1]; ++i1)
         {
-          A[(r * _bs[1] + i0) * ncols * _bs[0] + _cols[j] * _bs[1] + i1]
+          std::vector<std::int32_t> local_col {_cols[j]};
+          std::vector<std::int64_t> global_col{0};
+          _index_maps[1]->local_to_global(local_col, global_col);
```

Comment:
Couldn't we move this call outside the loop, apply `local_to_global` on all of `_cols` at the same time?
It would avoid having `std::vector<T>/std::array<T,1>` inside the loop (and avoid calling: https://github.com/FEniCS/dolfinx/blob/ce09b5e8b56cbd69bcf5f17afb4c2b7217b21706/cpp/dolfinx/common/IndexMap.cpp#L883) multiple times).

Reply:
That would work. But doesn't it reduce readability of the test case quite drastically? We would loose the clear correlation of for example here we look at the local index 4 but in terms of global indices.

Reply:
As long as we document it I think it is fine to add an initial comment stating:
```c++
// Transform all local column indices to global indices prior to looping through all rows
std::vector<std::int64_t> global_columns (_cols.size());
_index_maps[1]->local_to_global(_cols, global_columns);
```
and then in turn use `global_columns[j]` inside the loops.

Lets at least move the transformation out to line 637, as it is now called for every block size, which is unnecessary. And use `std::array<T, 1>` for the allocation.

Reply:
How about we use an `std::map` if we precompute? That way we keep the info in the line of the check?

Reply:
I think it becomes a rather expensive (map lookup) and not necessary.
This isn't the most performance critical part of the code, but we should avoid potentially expensive patterns when necessary.

I'll let some of the other developers chime in on this as well, maybe @chrisrichardson or @garth-wells has some other opinions than I.

Reply:
Let's bear in mind this is really for debugging, so performance is not too important.